File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/embed/MedicationDto.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto.embed;

import org.taktik.icure.services.external.rest.v1.dto.CodeDto;

import java.io.Serializable;
import java.util.List;
import java.util.Map;

public class MedicationDto implements Serializable {
	protected String compoundPrescription;
	protected SubstanceproductDto substanceProduct;
	protected MedicinalproductDto medicinalProduct;

	Integer numberOfPackages;

	String batch;

	String instructionForPatient;
	String commentForDelivery;
	String drugRoute; //CD-DRUG-ROUTE
	String temporality; //CD-TEMPORALITY : chronic, acute, oneshot

	DurationDto duration;
	RenewalDto renewal;

	Long beginMoment;
	Long endMoment;

	Boolean knownUsage;

	CodeDto frequency;
	CodeDto reimbursementReason;
	Boolean substitutionAllowed;


	List<RegimenItemDto> regimen;
	String posology;

	Map<String, ContentDto> options;
	Map<String, ParagraphAgreementDto> agreements;

	String medicationSchemeIdOnSafe;
	Integer medicationSchemeSafeVersion;
	Long medicationSchemeTimeStampOnSafe;
	String medicationSchemeDocumentId;
	String safeIdName; //can be: vitalinkuri, RSWID, RSBID
	String idOnSafes; //medicationschemeelement : value of vitalinkuri, RSBID, RSWID
	Long timestampOnSafe; //transaction date+time
	Boolean changeValidated; //accept change on safe
	Boolean newSafeMedication; //new medication on safe
	String medicationUse; //free text
	String beginCondition; //free text
	String endCondition; //free text
	String origin; // regularprocess, recorded
	Boolean medicationChanged;
	Boolean posologyChanged;

	String prescriptionRID;

	public Map<String, ContentDto> getOptions() {
		return options;
	}

	public void setOptions(Map<String, ContentDto> options) {
		this.options = options;
	}

	public String getCompoundPrescription() {
		return compoundPrescription;
	}

	public void setCompoundPrescription(String compoundPrescription) {
		this.compoundPrescription = compoundPrescription;
	}

	public SubstanceproductDto getSubstanceProduct() {
		return substanceProduct;
	}

	public void setSubstanceProduct(SubstanceproductDto substanceProduct) {
		this.substanceProduct = substanceProduct;
	}

	public MedicinalproductDto getMedicinalProduct() {
		return medicinalProduct;
	}

	public void setMedicinalProduct(MedicinalproductDto medicinalProduct) {
		this.medicinalProduct = medicinalProduct;
	}

	public Integer getNumberOfPackages() {
		return numberOfPackages;
	}

	public void setNumberOfPackages(Integer numberOfPackages) {
		this.numberOfPackages = numberOfPackages;
	}

	public String getBatch() {
		return batch;
	}

	public void setBatch(String batch) {
		this.batch = batch;
	}

	public Boolean getKnownUsage() {
		return knownUsage;
	}

	public void setKnownUsage(Boolean knownUsage) {
		this.knownUsage = knownUsage;
	}

	public String getInstructionForPatient() {
		return instructionForPatient;
	}

	public void setInstructionForPatient(String instructionForPatient) {
		this.instructionForPatient = instructionForPatient;
	}

	public String getCommentForDelivery() {
		return commentForDelivery;
	}

	public void setCommentForDelivery(String commentForDelivery) {
		this.commentForDelivery = commentForDelivery;
	}

	public Long getBeginMoment() {
		return beginMoment;
	}

	public void setBeginMoment(Long beginMoment) {
		this.beginMoment = beginMoment;
	}

	public Long getEndMoment() {
		return endMoment;
	}

	public void setEndMoment(Long endMoment) {
		this.endMoment = endMoment;
	}

	public List<RegimenItemDto> getRegimen() {
		return regimen;
	}

	public void setRegimen(List<RegimenItemDto> regimen) {
		this.regimen = regimen;
	}

	public String getPosology() { return posology; }

	public void setPosology(String posology) { this.posology = posology; }

	public DurationDto getDuration() {
		return duration;
	}

	public void setDuration(DurationDto duration) {
		this.duration = duration;
	}

	public Map<String, ParagraphAgreementDto> getAgreements() {
		return agreements;
	}

	public void setAgreements(Map<String, ParagraphAgreementDto> agreements) {
		this.agreements = agreements;
	}

	public CodeDto getFrequency() { return frequency; }

	public void setFrequency(CodeDto frequency) { this.frequency = frequency; }

	public CodeDto getReimbursementReason() { return reimbursementReason; }

	public void setReimbursementReason(CodeDto reimbursementReason) { this.reimbursementReason = reimbursementReason; }

	public Boolean getSubstitutionAllowed() { return substitutionAllowed; }

	public void setSubstitutionAllowed(Boolean substitutionAllowed) { this.substitutionAllowed = substitutionAllowed; }

	public RenewalDto getRenewal() { return renewal; }

	public void setRenewal(RenewalDto renewal) { this.renewal = renewal; }

	public String getDrugRoute() { return drugRoute; }

	public void setDrugRoute(String drugRoute) { this.drugRoute = drugRoute; }

	public String getTemporality() { return temporality; }

	public void setTemporality(String temporality) { this.temporality = temporality; }

	public String getMedicationSchemeIdOnSafe() { return medicationSchemeIdOnSafe; }

	public void setMedicationSchemeIdOnSafe(String medicationSchemeIdOnSafe) { this.medicationSchemeIdOnSafe = medicationSchemeIdOnSafe; }

	public Integer getMedicationSchemeSafeVersion() { return medicationSchemeSafeVersion; }

	public void setMedicationSchemeSafeVersion(Integer medicationSchemeSafeVersion) { this.medicationSchemeSafeVersion = medicationSchemeSafeVersion; }

	public Long getMedicationSchemeTimeStampOnSafe() { return medicationSchemeTimeStampOnSafe; }

	public void setMedicationSchemeTimeStampOnSafe(Long medicationSchemeTimeStampOnSafe) { this.medicationSchemeTimeStampOnSafe = medicationSchemeTimeStampOnSafe; }

	public String getMedicationSchemeDocumentId() { return medicationSchemeDocumentId; }

	public void setMedicationSchemeDocumentId(String medicationSchemeDocumentId) { this.medicationSchemeDocumentId = medicationSchemeDocumentId; }

	public String getSafeIdName() { return safeIdName; }

	public void setSafeIdName(String safeIdName) { this.safeIdName = safeIdName; }

	public String getIdOnSafes() { return idOnSafes; }

	public void setIdOnSafes(String idOnSafes) { this.idOnSafes = idOnSafes; }

	public Long getTimestampOnSafe() { return timestampOnSafe; }

	public void setTimestampOnSafe(Long timestampOnSafe) { this.timestampOnSafe = timestampOnSafe; }

	public Boolean getChangeValidated() { return changeValidated; }

	public void setChangeValidated(Boolean changeValidated) { this.changeValidated = changeValidated; }

	public Boolean getNewSafeMedication() { return newSafeMedication; }

	public void setNewSafeMedication(Boolean newSafeMedication) { this.newSafeMedication = newSafeMedication; }

	public String getMedicationUse() { return medicationUse; }

	public void setMedicationUse(String medicationUse) { this.medicationUse = medicationUse; }

	public String getBeginCondition() { return beginCondition; }

	public void setBeginCondition(String beginCondition) { this.beginCondition = beginCondition; }

	public String getEndCondition() { return endCondition; }

	public void setEndCondition(String endCondition) { this.endCondition = endCondition; }

	public String getOrigin() { return origin; }

	public void setOrigin(String origin) { this.origin = origin; }

	public Boolean getMedicationChanged() { return medicationChanged; }

	public void setMedicationChanged(Boolean medicationChanged) { this.medicationChanged = medicationChanged; }

	public Boolean getPosologyChanged() { return posologyChanged; }

	public void setPosologyChanged(Boolean posologyChanged) { this.posologyChanged = posologyChanged; }

	public String getPrescriptionRID() { return prescriptionRID; }

	public void setPrescriptionRID(String prescriptionRID) { this.prescriptionRID = prescriptionRID; }
}

File: src/main/kotlin/org/taktik/icure/services/external/rest/v1/dto/embed/MedicationDto.kt
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.rest.v1.dto.embed

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.github.pozo.KotlinBuilder
import org.taktik.icure.services.external.rest.v1.dto.base.CodeStubDto
import java.io.Serializable


@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class MedicationDto(
        val compoundPrescription: String? = null,
        val substanceProduct: SubstanceproductDto? = null,
        val medicinalProduct: MedicinalproductDto? = null,
        val numberOfPackages: Int? = null,
        val batch: String? = null,
        val instructionForPatient: String? = null,
        val commentForDelivery: String? = null,
        val drugRoute: String? = null, //CD-DRUG-ROUTE
        val temporality: String? = null, //CD-TEMPORALITY : chronic, acute, oneshot
        val frequency: CodeStubDto? = null, //CD-PERIODICITY
        val reimbursementReason: CodeStubDto? = null,
        val substitutionAllowed: Boolean? = null,
        val beginMoment: Long? = null,
        val endMoment: Long? = null,
        val deliveryMoment: Long? = null,
        val endExecutionMoment: Long? = null,
        val duration: DurationDto? = null,
        val renewal: RenewalDto? = null,
        val knownUsage: Boolean? = null,
        val regimen: List<RegimenItemDto>? = null,
        val posology: String? = null, // replace structured posology by text
        //val options: Map<String, ContentDto>? = null, Evil
        val agreements: Map<String, ParagraphAgreementDto>? = null,
        val medicationSchemeIdOnSafe: String? = null,
        val medicationSchemeSafeVersion: Int? = null,
        val medicationSchemeTimeStampOnSafe: Long? = null,
        val medicationSchemeDocumentId: String? = null,
        val safeIdName: String? = null, //can be: vitalinkuri, RSWID, RSBID
        val idOnSafes: String? = null, //medicationschemeelement : value of vitalinkuri, RSBID, RSWID
        val timestampOnSafe: Long? = null, //transaction date+time
        val changeValidated: Boolean? = null, //accept change on safe
        val newSafeMedication: Boolean? = null, //new medication on safe
        val medicationUse: String? = null, //free text
        val beginCondition: String? = null, //free text
        val endCondition: String? = null, //free text
        val origin: String? = null, // regularprocess, recorded
        val medicationChanged: Boolean? = null,
        val posologyChanged: Boolean? = null,
        val suspension: List<SuspensionDto>? = null,
        val prescriptionRID: String? = null,
        val status: Int? = null
) : Serializable {
    companion object {
        const val REIMBURSED = "REIMBURSED"
        const val STATUS_NOT_SENT = 1 shl 0 //not send by recip-e
        const val STATUS_SENT = 1 shl 1 //sent by recip-e
        const val STATUS_PENDING = 1 shl 2 //not delivered to patient
        const val STATUS_DELIVERED = 1 shl 3 //delivered to patient
        const val STATUS_REVOKED = 1 shl 4 //revoked by physician
    }
}
